### PR TITLE
QoL code improvements for dashboard

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
The vast majority of these should affect code style only.
This PR is intended to be quick to code review + QA - if something doesn't belong here it will be easy enough to move it to another PR.
One bug in the development mode service worker (`app/ide-desktop/lib/content/src/devServiceWorker.ts`) is fixed.

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

**Review by wdanilo — APPROVED** [review]

File: app/ide-desktop/eslint.config.js [hunk 1]
```diff
@@ -347,6 +354,7 @@ export default [
             ],
             '@typescript-eslint/restrict-template-expressions': 'error',
             '@typescript-eslint/sort-type-constituents': 'error',
+            // '@typescript-eslint/strict-boolean-expressions': 'error',
```
Comment (indiv0) [inline app/ide-desktop/eslint.config.js:357]: Why are we adding a commented rule?

File: app/ide-desktop/lib/client/watch.ts [hunk 2]
```diff
@@ -107,6 +109,10 @@ const ALL_BUNDLES_READY = new Promise<Watches>((resolve, reject) => {
                 })
             },
         })
+        contentOpts.pure.splice(contentOpts.pure.indexOf('assert'), 1)
+        ;(contentOpts.inject = contentOpts.inject ?? []).push(
```
Comment (indiv0) [inline app/ide-desktop/lib/client/watch.ts:113]: Typo with this semicolon?

File: app/ide-desktop/lib/content/src/index.ts [hunk 3]
```diff
@@ -37,7 +38,7 @@ const FETCH_TIMEOUT = 300
 // === Live reload ===
 // ===================
 
-if (IS_DEV_MODE) {
+if (IS_DEV_MODE && !detect.isRunningInElectron()) {
```
Comment (indiv0) [inline app/ide-desktop/lib/content/src/index.ts:41]: Why are we disabling live reload when running in Electron?

File: app/ide-desktop/lib/content/watch.ts [hunk 4]
```diff
@@ -38,6 +38,8 @@ async function watch() {
             supportsDeepLinks: false,
         })
     )
+    opts.pure.splice(opts.pure.indexOf('assert'), 1)
+    ;(opts.inject = opts.inject ?? []).push(path.resolve(THIS_PATH, '..', '..', 'debugGlobals.ts'))
```
Comment (indiv0) [inline app/ide-desktop/lib/content/watch.ts:42]: Typo with this semicolon?

File: app/ide-desktop/lib/content/src/newtype.ts [hunk 5]
```diff
@@ -1,46 +0,0 @@
-/** @file Emulates `newtype`s in TypeScript. */
```
Comment (indiv0) [inline app/ide-desktop/lib/content/src/newtype.ts:1]: Is this file unused now?

**Review by indiv0 — CHANGES_REQUESTED** [review]

**Comment by indiv0** [comment]
**TL;DR**: QA complete, some non-issues and some potential issues found.

# Scenario Descriptions

<table style="border: 1px solid black;">
    <thead>
        <tr>
            <th style="border: 1px solid black;" rowspan=4>Scenario</th>
        </tr>
        <tr>
            <th style="border: 1px solid black;" colspan=5>Command</th>
            <th style="border: 1px solid black;" colspan=2>Features</th>
            <th style="border: 1px solid black;" rowspan=3>Debug Scene</th>
        </tr>
        <tr style="border: 1px solid black;">
            <th style="border: 1px solid black;">GUI</th>
            <th style="border: 1px solid black;" colspan=2>IDE</th>
            <th style="border: 1px solid black;" colspan=2>NPM</th>
            <th style="border: 1px solid black;" rowspan=2>Auth</th>
            <th style="border: 1px solid black;" rowspan=2>New Dashboard</th>
        </tr>
        <tr style="border: 1px solid black;">
            <th style="border: 1px solid black;">Watch</th>
            <th style="border: 1px solid black;">Watch</th>
            <th style="border: 1px solid black;">Build</th>
            <th style="border: 1px solid black;">build-dashboard</th>
            <th style="border: 1px solid black;">watch-dashboard</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td style="border: 1px solid black;">0</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">❌</td>
            <td style="border: 1px solid black;">❌</td>
            <td style="border: 1px solid black;">❌</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">1</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">2</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">❌</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">3</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">❌</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">4</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
        </tr>
        <tr>
            <td style="border: 1px solid black;">5</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">✅</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
            <td style="border: 1px solid black;">-</td>
        </tr>
    </tbody>
</table>

# Scenario 0

```sh
./run gui watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info
```

- ✅ GUI at http://localhost:8080 opens
- ✅ Can login
- ✅ Dashboard shows up
- ❌ ~~Potential issue w/ DOM validation, see screenshot~~
  - Known issue, not introduced in this PR, fixed in a separate PR per [this discussion](https://github.com/enso-org/enso/pull/7022#issuecomment-1594928048)
- ✅ Project manager doesn't connect, see screenshot (though I'm not sure, is this even supposed to work when connecting via http://localhost:8080 ?)
  - Working as intended, `./run gui watch` does not run the project manager per [this discussion](https://github.com/enso-org/enso/pull/7022#issuecomment-1594924403)
- ✅ IDE rebuilds for edits in `app/gui/**/*.rs`
- ✅ IDE reloads for edits in `app/gui/**/*.rs`
- ✅ IDE rebuilds for edits in `app/ide-desktop/lib/dashboard/src/authentication/src/**/*.tsx`
- ✅ IDE reloads for edits in `app/ide-desktop/lib/dashboard/src/authentication/src/**/*.tsx`


<details>
<summary>
Screenshot
</summary>
image
</details>

# Scenario 1

```sh
ENSO_BUILD_TARGET=dir \
    ./run ide watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --backend-source=release \
    --backend-release=nightly \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info \
    --ide-option=-feature-preview.new-dashboard \
    --ide-option=-debug.dev-tools \
    --ide-option=-debug \
    --ide-option=--inspect=5858 \
    --ide-option=-authentication \
    --ide-option=-startup.entry=text_area
```

- ✅ Debug scene is visible
- ✅ Logs appear with `console.log("Hello")`

<details>
<summary>
Screenshot
</summary>
image

</details>

# Scenario 2

```sh
ENSO_BUILD_TARGET=dir \
    ./run ide watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --backend-source=release \
    --backend-release=nightly \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info \
    --ide-option=-feature-preview.new-dashboard \
    --ide-option=-debug.dev-tools \
    --ide-option=-debug \
    --ide-option=--inspect=5858 \
    --ide-option=-authentication
```

- ✅ Nodes in Cloud project are visible
- ❌ ~~Potential issue w/ DOM validation, see screenshot~~
  - Known issue, not introduced in this PR, fixed in a separate PR per [this discussion](https://github.com/enso-org/enso/pull/7022#issuecomment-1594928048)
- ✅ IDE rebuilds for edits in `app/gui/**/*.rs`
- ✅ IDE does not reload for edits in `app/gui/**/*.rs`
  - Expected behaviour per [this discussion](https://github.com/enso-org/enso/pull/7022#discussion_r1232442430)
- ✅ IDE rebuilds for edits in `app/ide-desktop/lib/dashboard/src/authentication/src/**/*.tsx`
- ✅ IDE does not reload for edits in `app/ide-desktop/lib/dashboard/src/authentication/src/**/*.tsx`
  - Expected behaviour per [this discussion](https://github.com/enso-org/enso/pull/7022#discussion_r1232442430)

<details>
<summary>
Screenshot
</summary>
image
</details>

# Scenario 3

```sh
ENSO_BUILD_TARGET=dir \
    ./run ide build \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --backend-source=release \
    --backend-release=nightly \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info
./dist/ide/mac-arm64/Enso.app/Contents/MacOS/Enso \
    -feature-preview.new-dashboard \
    -debug.dev-tools \
    -debug \
    --inspect=5858 \
    -authentication
```

- ✅ Nodes in Cloud project are visible

<details>
<summary>
Screenshot
</summary>
image

</details>

# Scenario 4

```sh
cd app/ide-desktop
npm run build-dashboard
```

- ✅ Command ran successfully

<details>
<summary>
Screenshot
</summary>
image

</details>




# Scenario 5

```sh
cd app/ide-desktop
npm run watch-dashboard
```

- ✅ Command ran successfully
- ✅ GUI at http://localhost:8080 opens
- ❌ ~~Potential issue w/ DOM validation, see screenshot~~
  - Known issue, not introduced in this PR, fixed in a separate PR per [this discussion](https://github.com/enso-org/enso/pull/7022#issuecomment-1594928048)
- ❌ ~~Potential issue w/ `serviceWorker.js`, see screenshot 5.1~~
  - Not an issue, red-herring that will be fixed in this PR, per [this discussion](https://github.com/enso-org/enso/pull/7022#issuecomment-1595049692)
- ✅ IDE does not rebuild for edits in `app/gui/**/*.rs`
- ✅ IDE rebuilds for edits in `app/ide-desktop/lib/dashboard/src/authentication/src/**/*.tsx`
- ✅ IDE reloads for edits in `app/ide-desktop/lib/dashboard/src/authentication/src/**/*.tsx`
- ❌ ~~Potential issue w/ images for templates not loading (e.g. `http://localhost:8080/visualize.png`), see screenshot 5.2~~
  - Known issue, not introduced in this PR, per [this discussion](https://github.com/enso-org/enso/pull/7022#issuecomment-1594968308)
- ❌ ~~Potential issue w/ images in project (e.g. `http://localhost:8080/spreadsheets.png`), see screenshot 5.3~~
  - Known issue, not introduced in this PR, per [this discussion](https://github.com/enso-org/enso/pull/7022#issuecomment-1594968308)
- ✅ Nodes in Cloud project are visible

<details>
<summary>
Screenshots
</summary>
image
image
image

</details>

File: app/ide-desktop/lib/content/watch.ts (hunk 4 above)
Comment (indiv0) [reply app/ide-desktop/lib/content/watch.ts:42]: JS thinks this is a function call without semicolon, so this is OK.

File: app/ide-desktop/lib/client/watch.ts (hunk 2 above)
Comment (indiv0) [reply app/ide-desktop/lib/client/watch.ts:113]: JS thinks this is a function call without semicolon, so this is OK.

File: app/ide-desktop/lib/content/src/newtype.ts (hunk 5 above)
Comment (indiv0) [reply app/ide-desktop/lib/content/src/newtype.ts:1]: Per @somebody1234:
> it was used in project manager code which was moved to the dashboard

File: app/ide-desktop/eslint.config.js (hunk 1 above)
Comment (indiv0) [reply app/ide-desktop/eslint.config.js:357]: Per @somebody1234:
> commented rule: will be uncommented in a future PR, feel free to remove

File: app/ide-desktop/lib/content/src/index.ts (hunk 3 above)
Comment (indiv0) [reply app/ide-desktop/lib/content/src/index.ts:41]: Per @somebody1234:
> why we are disabling live reload in electron: live reload depends on esbuild server, electron app doesn't run the esbuild server (main issue is that the ports conflict). it might be possible to run esbuild from inside electron, it might be a good idea to open an issue on the cloud-v2 repo for that?
>
> yeah, i dont think it's worked for a while - not since i switched it to esbuild at least
note that devmode was previously false for ide watch: https://github.com/enso-org/enso/pull/7022/files#diff-630face3223d6e85ea83529c29722ede4d3573044b4fd33590a778b26b455346R100
> so the live reload wouldn't have been enabled anyway

**Comment by somebody1234** [comment]
project manager doesn't connect: working as intended - gui watch does not run the project manager. of course, you can manually run the project manager binary separately.
(ide watch runs the project manager because the electron entrypoint starts the project manager)

**Comment by somebody1234** [comment]
dom validation issue: known issue, not introduced in this PR, already being fixed in another PR

File: app/ide-desktop/lib/content/src/index.ts (hunk 3 above)
Comment (somebody1234) [reply app/ide-desktop/lib/content/src/index.ts:41]: crosspost from discord: live reload relies on the `/esbuild` endpoint from esbuild, which does not exist when running the electron app - note that this PR also changes `IS_DEV_MODE` from `false` to `true` for ide watch

**Comment by somebody1234** [comment]
assets not loading: known issue (has always been an issue). not sure there's a satisfactory solution, but it might be good to open up an issue so potential solutions can be discussed.

either way it shouldn't be a regression at the very least

**Comment by somebody1234** [comment]
re: `failed to load ''` - i don't think i've seen that before, could you possibly find it in the network tab?

**Comment by indiv0** [comment]
> re: `failed to load ''` - i don't think i've seen that before, could you possibly find it in the network tab?

Error seems to happen between the request for `./run.js` and the request for `/favicon.ico` so from the requests tab it looks like it might be the request for `/esbuild` that's failing?

image

**Comment by somebody1234** [comment]
weird... i'm getting completely different errors on `watch-dashboard`... it'd probably be nice if you manage to find the cause, but i don't think it's too urgent - there are other PRs needing QA after all
on an unrelated note it's pretty late here so i'll be sleeping soon

**Comment by indiv0** [comment]
This might be a firefox thing? I'll see if I get the same errors in Chrome. For what it's worth, the line pointed to by that error is this:
image

I'm happy to help debug, but I'm not sure how to. `console.log` doesn't seem to work within the `serviceWorker`. Not sure if that's a serviceWorker-specific issue or because the fixes for logging that were merged to `develop` aren't present on this branch.

**Comment by somebody1234** [comment]
actually it might possibly be `run.js` having an issue since it says `ns_error_corrupted_content`. not quite sure what could cause that though...

**Comment by indiv0** [comment]
I think that might be a red-herring because `run.js` is returning a 404 with `text/plain` content but since it's supposed to be a JavaScript file then `text/plain` is incorrect, hence the `NS_ERROR_CORRUPTED_CONTENT`. More relevant error:

image

**Comment by somebody1234** [comment]
maybe - note that `run.js` (along with some of the other stylesheets) can't be removed from `index.html` because it's required for the electron server/entrypoint

also worth noting `watch-dashboard` shouldn't be loading `run.js` anyway (but as above, it can't be removed) so i *think* currently just leaving it like that is the best option

**Comment by indiv0** [comment]
Looks like the service worker error is caused by the `return false` at the end of the listener. Changing it to just `return` gets rid of the service worker error in FF. So perhaps `return false` just isn't the right way to exit a service worker?

I'll wait for confirmation, but it seems like this is not really an issue, so we can count this as "QA approved".

```typescript
self.addEventListener('fetch', event => {
    const url = new URL(event.request.url)
    if (url.hostname === 'localhost' && url.pathname !== '/esbuild') {
        const responsePromise = /\/[^.]+$/.test(new URL(event.request.url).pathname)
            ? fetch('/index.html')
            : fetch(event.request.url)
        event.respondWith(
            responsePromise.then(response => {
                const clonedResponse = new Response(response.body, response)
                for (const [header, value] of common.COOP_COEP_CORP_HEADERS) {
                    clonedResponse.headers.set(header, value)
                }
                return clonedResponse
            })
        )
        return
    } else {
        //return false // <--- this line is the culprit
        return
    }
})
```

**Review by indiv0 — APPROVED** [review]
Approved, all comments addressed. ~~Awaiting final response RE: service worker issue discovered in QA.~~

Service worker issue fixed, ready for merge, pending CI checks.

**Comment by somebody1234** [comment]
yeah, not sure why i decided to return false but it's definitely not there for some special reason. it doesn't seem to break when i change `return false` to `return` locally so :+1: i think it's fine

**Comment by indiv0** [comment]
CI failure appears to be spurious, PR should still be good to merge:
```
ERROR main_internal:Upload worker{index=0}:upload_file{local_path=/Users/runner/work/enso/enso/engine/runtime/bench-report.xml remote_path=Runtime Benchmark Report/bench-report.xml upload_url=https://pipelines.actions.githubusercontent.com/UsybhwCQhLa83Psn06RntgFfGy6jLF6HygBizIhSvyZMcDTapY/_apis/resources/Containers/34454524}: ide_ci::actions::artifacts::raw: error=Failed to upload the file '/Users/runner/work/enso/enso/engine/runtime/bench-report.xml' to path 'Runtime Benchmark Report/bench-report.xml'.
```

Maybe re-run the CI?
